ボウリングのプログラムを提出いたします

ham-cap opened the pull request:
ボウリングのスコア計算プログラムを提出いたします。
ご確認の程よろしくお願いいたします。

ham-cap committed:
ストライクの判定を修正したもの。ゲームの最後の投球とその前の投球をnextで無視するようにしてしまっているため要修正。

ham-cap committed:
必要なときにも最後の2フレームを無視してしまうのを修正。

ham-cap committed:
10フレーム目でストライクを出した際、存在しないフレームを参照してしまっていたのを修正。

ham-cap committed:
rubocopによる修正を反映

インデントや空白文字の有無をrubocopに指摘されたため修正した。

maedana commented on 04.bowling/bowling.rb:
```diff
@@ -0,0 +1,46 @@
+  elsif index == frames.size - 2 && frames.size - 2 < 9
+    points += frame.sum
+  elsif index == frames.size - 1 || index == frames.size - 2
+    next
```
これより上は全部最終フレーム用の処理ですよね。ここ実は凄くシンプルに出来ます。

というのも、よく考えてみて欲しいのですが、最終フレームって実は全部足せばいいだけなんです。

```
X,X,X => 10 + 10 + 10
6,4,5 => 6 + 4 + 5
X, 4, 5 => 10 + 4 + 5
0, 4 => 0 + 4
```
のような。

後は今のロジックですとフレーム分割が単純な方法なので、最大12になってしまっているわけですよね。つまりループが最大で12回回ってしまうと。なので`最終フレームのとき`、`最終フレームを超えたとき`の2分岐があれば十分じゃないかなーと思うんです。いかがでしょうか?

ham-cap committed:
10フレーム目の処理を簡潔にした

10フレーム目については1〜3投目を単純に合計すればよいだけなので
インデックス番号が9以上のフレームを合計して先に変数pointsへ
代入しておき、以降のelsifで1〜9フレーム目の処理をすることとした。

ham-cap committed:
rubocopによる修正を反映

ham-cap commented:
いただいたアドバイスに従って10フレーム目の処理を書き直してみましたのでご確認願います。
10フレーム目は単純に足し算するだけというのは分かってはいたのですが、難しく考えすぎてどう書いていいやらわからなかったというのが正直なところです💧
`最終フレーム`と`最終フレーム以降`の分岐で十分なのではというご指摘の意図を反映したつもりなのですが、いかがでしょうか。
余談ですが、最初はcase文を使って10フレーム目の処理を書いて、1〜9フレーム目までの処理は元のif文のままcase分のelseの中に埋め込むという書き方をしてみて上手く動いたのですが、よく考えたらif文のままでもいけるなと気づいたので結局今回の修正になりました。
だいぶ短くなりましたし、ちゃんと動いていると思いますが、うまく修正できていますでしょうか？

maedana commented on 04.bowling/bowling.rb:
```diff
@@ -0,0 +1,34 @@
+end
+
+frames = []
+shots.each_slice(2) { |s| frames << s }
```
ご参考までに。以下のように書けばいきなり代入できますよ
```ruby
frames = shots.each_slice(2).to_a
```

maedana commented on 04.bowling/bowling.rb:
```diff
@@ -0,0 +1,34 @@
+points = 0
+
+frames.each_with_index do |frame, index|
+  points += if index >= 9
```
`point +=`を外に出したんですね。分岐が簡略化されてから直してもらおうかなと思っていた部分です。とてもよいと思います :+1: 

せっかくなので、さらにいうとここもframesのようにいきなり代入する形に出来たりします。

https://docs.ruby-lang.org/ja/latest/method/Array/i/sum.html

というのがあり、今も`frame.sum`で使っていると思いますが、配列の各要素についてブロックの実行結果を合計という使い方も出来るのです。

```ruby
points =
  frames.each_with_index.sum do |frame, index|
     # ここの実行結果が合計される
  end
```
ということです。伝わりますでしょうか。

maedana replied:
分岐についても凄くシンプルになりましたね。

> 10フレーム目は単純に足し算するだけというのは分かってはいたのですが、難しく考えすぎてどう書いていいやらわからなかったというのが正直なところです

なるほど、時間おいて改めてやってみたらあっさり出来たという感じですかね。正直もっと時間がかかるのかなと思っていたんですが、すぐにシンプルな答えを出してきたので素晴らしいなと思いました :+1:  かなり色々考えた経験が生きているのかなと思いましたー。

ちなみに僕が考えていたよりシンプルです :smile: 僕はもうちょっと難しく考えてしまっていて、
```ruby
    if index > 9
      0
    elsif index == 9
      frame.sum + (frames[index + 1]&.sum || 0) + (frames[index + 2]&.sum || 0)
    elsif frame[0] == 10 && frames[index + 1][0] != 10
       ... 以下略
```
という感じでした。けど、@ham-cap の修正内容の方がシンプルでいいですね :+1:

ham-cap replied:
ありがとうございます！
めちゃくちゃ嬉しいです🔆

繰り返し処理をいきなり代入すると毎回上書きされてしまうのかと思っていました😲


> 時間おいて改めてやってみたらあっさり出来たという感じですかね。

まさにそんな感じです！自分でももっと時間かかると思っていたんですが、ヒントいただいた後に試しにやってみたらすんなりできてちょっと焦りました。
やはり誰かに見ていただくのは大事ですね😅

maedana replied:
> まさにそんな感じです！自分でももっと時間かかると思っていたんですが、ヒントいただいた後に試しにやってみたらすんなりできてちょっと焦りました。

プログラミングってそういうことよくあると思います。ただ、それには十分に悩んでよく考えることが前提かもしれません。今回日報からもよーく考え抜いたことが伝わりましたし、その結果スッと出てきた答えなのかなと思います :smile:
